unified_planning/engines/sequential_simulator.py

```diff
@@ -0,0 +1,210 @@
+
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import Callable, Dict, Iterator, List, Optional, Tuple, Union, cast
+import unified_planning as up
+from unified_planning.engines.compilers import Grounder
+from unified_planning.engines.engine import Engine
+from unified_planning.engines.mixins.simulator import Event, SimulatorMixin
+from unified_planning.exceptions import UPUsageError
+from unified_planning.plans import ActionInstance
+from unified_planning.walkers import StateEvaluator
+
+
+class InstantaneousEvent(Event):
+    '''Implements the Event class for an Instantaneous Action.'''
+
+    def __init__(self,
+                 conditions: List['up.model.FNode'],
+                 effects: List['up.model.Effect'],
+                 simulated_effect: Optional['up.model.SimulatedEffect'] = None):
+        self._conditions = conditions
+        self._effects = effects
+        self._simulated_effect = simulated_effect
+
+    @property
+    def conditions(self) -> List['up.model.FNode']:
+        return self._conditions
+
+    @property
+    def effects(self) -> List['up.model.Effect']:
+        return self._effects
+
+    @property
+    def simulated_effect(self) -> Optional['up.model.SimulatedEffect']:
+        return self._simulated_effect
+
+
+class SequentialSimulator(Engine, SimulatorMixin):
+    '''
+    Sequential SimulatorMixin implementation.
+    '''
+
+    def __init__(self, problem: 'up.model.Problem'):
+        SimulatorMixin.__init__(self, problem)
+        pk = problem.kind
+        if not SequentialSimulator.supports(pk):
+            raise UPUsageError(f'The problem named: {problem.name} is not supported by the SequentialSimulator.')
+        assert Grounder.supports(pk)
+        assert isinstance(self._problem, up.model.Problem)
+        grounder = Grounder()
+        self._grounding_result = grounder.compile(self._problem, up.engines.CompilationKind.GROUNDING)
+
+        grounded_problem: 'up.model.Problem' = cast(up.model.Problem, self._grounding_result.problem)
+        map = cast(Callable[[ActionInstance], ActionInstance], self._grounding_result.map_back_action_instance)
```

Comment:
`map` is a bad name, because it is a reserved keyword in Python. It works because you can redefine even built-ins, but it is a bad practice